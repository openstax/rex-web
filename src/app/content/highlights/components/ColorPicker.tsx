import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import React from 'react';
import { useIntl } from 'react-intl';
import styled from 'styled-components/macro';
import { match, not } from '../../../fpUtils';
import { highlightStyles } from '../../constants';
import { cardPadding } from '../constants';
import ColorIndicator from './ColorIndicator';

interface SingleSelectProps {
  color?: HighlightColorEnum;
  onRemove?: () => void;
  onChange: (color: HighlightColorEnum) => void;
  multiple: false | undefined;
}

interface MultipleSelectProps {
  selected: HighlightColorEnum[];
  onChange: (selected: string[]) => void;
  multiple: true;
}

type Props = (SingleSelectProps | MultipleSelectProps) & {
  className?: string;
  size?: 'small';
};

interface ColorButtonProps {
  name: string;
  checked: boolean;
  size: Props['size'];
  style: typeof highlightStyles[number];
  onClick: () => void;
  className?: string;
}

// tslint:disable-next-line:variable-name
const ColorButton = styled(({className, size, style, ...props}: ColorButtonProps) => {
  const color = useIntl().formatMessage({id: `i18n:highlighting:colors:${style.label}`});

  return <ColorIndicator
    style={style}
    size={size}
    title={color}
    aria-label={useIntl().formatMessage({id: 'i18n:highlighting:change-color'}, {color})}
    checked={props.checked}
    component={<label />}
    className={className}
  >
    <input type='checkbox' {...props} />
  </ColorIndicator>;
})`
  cursor: pointer;
  margin: 0 ${cardPadding}rem ${cardPadding}rem 0;

  input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
    height: 0;
    width: 0;
  }
`;

// tslint:disable-next-line:variable-name
const ColorPicker = ({className, ...props}: Props) => {

  // its important that this be click focusable, because when clicking
  // labels in chrome the focus is first moved to a background element
  // before the input, and that might cause weird behavior in parent
  // elements.
  return <div className={className} tabIndex={-1}>
    {highlightStyles.map((style) => <ColorButton key={style.label}
      name={style.label}
      checked={props.multiple ? props.selected.includes(style.label) : props.color === style.label}
      style={style}
      size={props.size}
      onChange={() => props.multiple
        ? props.selected.includes(style.label)
          ? props.onChange(props.selected.filter(not(match(style.label))))
          : props.onChange([...props.selected, style.label])
        : props.color === style.label
          ? props.onRemove ? props.onRemove() : null
          : props.onChange(style.label)}
    />)}
  </div>;
};

export default styled(ColorPicker)`
  outline: none;
  display: flex;
  flex-direction: row;
  overflow: visible;
`;
