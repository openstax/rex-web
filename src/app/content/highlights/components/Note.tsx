import { HTMLTextAreaElement } from '@openstax/types/lib.dom';
import React from 'react';
import { useIntl } from 'react-intl';
import styled from 'styled-components/macro';
import { textStyle } from '../../../components/Typography/base';
import theme from '../../../theme';
import { cardPadding, cardWidth } from '../constants';

interface Props {
  note: string;
  textareaRef: React.RefObject<HTMLTextAreaElement>;
  onChange: (note: string) => void;
  onFocus: () => void;
}

const noteMaxLength = 1000;

const width = cardWidth - cardPadding * 2;
// tslint:disable-next-line:variable-name
const TextArea = styled.textarea`
  display: block;
  min-height: 5.6rem;
  width: ${width}rem;
  max-height: 30rem;
  max-width: ${width}rem;
  min-width: ${width}rem;
  border: 1px solid ${theme.color.neutral.formBorder};
  padding: ${cardPadding}rem;
  ${textStyle}
  color: ${theme.color.text.label};
  font-size: 1.4rem;
  font-family: inherit;
  line-height: 2rem;
  font-weight: normal;
`;

// tslint:disable-next-line:variable-name
const Note = ({onChange, onFocus, note, textareaRef}: Props) => {
  const setTextAreaHeight = React.useCallback(() => {
    const element = textareaRef.current;
    if (!element) {
      return;
    }

    if (element.scrollHeight > element.offsetHeight) {
      element.style.height = `${element.scrollHeight + 5}px`;
    }
  }, [textareaRef]);

  React.useEffect(setTextAreaHeight, [note, setTextAreaHeight]);

  return <TextArea
    ref={textareaRef}
    value={note}
    onFocus={onFocus}
    maxLength={noteMaxLength}
    onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => {
      onChange(e.target.value);
    }}
    placeholder={useIntl().formatMessage({id: 'i18n:highlighting:card:placeholder'})}
  />;
};

export default Note;
