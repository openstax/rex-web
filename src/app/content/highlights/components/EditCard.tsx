import { Highlight } from '@openstax/highlighter';
import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import { HTMLElement, HTMLTextAreaElement } from '@openstax/types/lib.dom';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';
import styled, { css } from 'styled-components/macro';
import { useAnalyticsEvent } from '../../../../helpers/analytics';
import * as selectAuth from '../../../auth/selectors';
import Button, { ButtonGroup } from '../../../components/Button';
import { useFocusElement, useOnEsc } from '../../../reactUtils';
import theme from '../../../theme';
import { assertDefined, assertWindow, mergeRefs } from '../../../utils';
import { highlightStyles } from '../../constants';
import {
  clearFocusedHighlight,
  setAnnotationChangesPending as setAnnotationChangesPendingAction,
  updateHighlight,
} from '../actions';
import { cardPadding } from '../constants';
import { HighlightData } from '../types';
import { generateUpdatePayload } from './cardUtils';
import ColorPicker from './ColorPicker';
import Confirmation from './Confirmation';
import Note from './Note';
import { isElementForOnClickOutside, useOnClickOutside } from './utils/onClickOutside';

export interface EditCardProps {
  isActive: boolean;
  hasUnsavedHighlight: boolean;
  highlight: Highlight;
  locationFilterId: string;
  pageId: string;
  onCreate: () => void;
  onBlur: typeof clearFocusedHighlight;
  setAnnotationChangesPending: typeof setAnnotationChangesPendingAction;
  onRemove: () => void;
  onCancel: () => void;
  onHeightChange: (ref: React.RefObject<HTMLElement>) => void;
  data?: HighlightData;
  className: string;
  shouldFocusCard: boolean;
}

// tslint:disable-next-line:variable-name
const EditCard = React.forwardRef<HTMLElement, EditCardProps>((props, ref) => {
  const authenticated = !!useSelector(selectAuth.user);
  const loginLink = useSelector(selectAuth.loginLink);
  const dispatch = useDispatch();
  const defaultAnnotation = () => props.data && props.data.annotation ? props.data.annotation : '';
  const [pendingAnnotation, setPendingAnnotation] = React.useState<string>(defaultAnnotation());
  const [editingAnnotation, setEditing] = React.useState<boolean>(!!props.data && !!props.data.annotation);
  const [confirmingDelete, setConfirmingDelete] = React.useState<boolean>(false);
  const element = React.useRef<HTMLElement>(null);
  const textarea = React.useRef<HTMLTextAreaElement>(null);

  const trackCreateNote = useAnalyticsEvent('createNote');
  const trackEditNoteColor = useAnalyticsEvent('editNoteColor');
  const trackEditAnnotation = useAnalyticsEvent('editAnnotation');
  const trackShowCreate = useAnalyticsEvent('showCreate');
  const trackShowLogin = useAnalyticsEvent('showLogin');
  const trackDeleteHighlight = useAnalyticsEvent('deleteHighlight');

  const blurIfNotEditing = React.useCallback(() => {
    if (!props.hasUnsavedHighlight && !editingAnnotation) {
      props.onBlur();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props.hasUnsavedHighlight, editingAnnotation]);

  const cancelEditing = () => {
    setPendingAnnotation(defaultAnnotation());
    props.setAnnotationChangesPending(false);
    setEditing(false);
    props.onCancel();
  };

  useOnEsc(element, props.isActive, cancelEditing);

  React.useEffect(() => {
    if (props.data) { return; }
    if (authenticated) {
      trackShowCreate();
    } else {
      trackShowLogin();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const elements = React.useMemo(
    () => [element, ...props.highlight.elements].filter(isElementForOnClickOutside),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [element.current, props.highlight]);

  useOnClickOutside(elements, props.isActive, blurIfNotEditing, { capture: true });

  React.useEffect(() => {
    if (element.current) {
      props.onHeightChange(element);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [element, editingAnnotation, props.isActive]);

  useFocusElement(textarea, props.shouldFocusCard);

  const onColorChange = (color: HighlightColorEnum, isDefault?: boolean) => {
    props.highlight.setStyle(color);
    if (props.data) {
      const {updatePayload, preUpdateData} = generateUpdatePayload(props.data, {color, id: props.data.id});

      dispatch(updateHighlight(updatePayload, {
        locationFilterId: props.locationFilterId,
        pageId: props.pageId,
        preUpdateData,
      }));
      trackEditNoteColor(color);
    } else {
      assertWindow().getSelection()?.removeAllRanges();
      props.onCreate();
      trackCreateNote(isDefault ? 'default' : color);
    }
  };

  const saveAnnotation = (toSave: HighlightData) => {
    const data  = assertDefined(props.data, 'Can\'t update highlight that doesn\'t exist');

    const addedNote = data.annotation === undefined;
    const {updatePayload, preUpdateData} =
      generateUpdatePayload(data, {id: toSave.id, annotation: pendingAnnotation});

    dispatch(updateHighlight(updatePayload, {
      locationFilterId: props.locationFilterId,
      pageId: props.pageId,
      preUpdateData,
    }));
    trackEditAnnotation(addedNote, toSave.color);
    props.onCancel();
    props.highlight.focus();
  };

  const updateUnsavedHighlightStatus = (newValue: string) => {
    const currentValue = props.data && props.data.annotation ? props.data.annotation : '';
    if (currentValue !== newValue && !props.hasUnsavedHighlight) {
      props.setAnnotationChangesPending(true);
    }

    if (currentValue === newValue && props.hasUnsavedHighlight) {
      props.setAnnotationChangesPending(false);
    }
  };

  return <form
    className={props.className}
    ref={mergeRefs(ref, element)}
    data-analytics-region='edit-note'
    data-highlight-card
  >
    <ColorPicker color={props.data ? props.data.color : undefined} onChange={onColorChange} onRemove={() => {
      if (props.data && !props.data.annotation && !pendingAnnotation) {
        props.onRemove();
        trackDeleteHighlight(props.data.color);
      }
    }} />
    <Note
      textareaRef={textarea}
      note={pendingAnnotation}
      onFocus={() => {
        if (!props.highlight.getStyle()) {
          onColorChange(highlightStyles[0].label, true);
        }
      }}
      onChange={(newValue) => {
        setPendingAnnotation(newValue);
        updateUnsavedHighlightStatus(newValue);
        setEditing(true);
      }}
    />
    {editingAnnotation && props.data && <ButtonGroup>
      <FormattedMessage id='i18n:highlighting:button:save'>
        {(msg) => <Button
          data-testid='save'
          data-analytics-label='save'
          size='small'
          variant='primary'
          onClick={(e: React.FormEvent) => {
            e.preventDefault();
            setEditing(false);
            const data = assertDefined(props.data, 'props.data should be defined');
            if (pendingAnnotation === '' && data.annotation) {
              setConfirmingDelete(true);
            } else {
              saveAnnotation(data);
            }
          }}
        >{msg}</Button>}
      </FormattedMessage>
      <FormattedMessage id='i18n:highlighting:button:cancel'>
        {(msg) => <Button
          size='small'
          data-analytics-label='cancel'
          data-testid='cancel'
          onClick={(e: React.FormEvent) => {
            e.preventDefault();
            cancelEditing();
          }}
        >{msg}</Button>}
      </FormattedMessage>
    </ButtonGroup>}
    {confirmingDelete && props.data && <Confirmation
      data-testid='confirm-delete'
      data-analytics-region='highlighting-delete-note'
      message='i18n:highlighting:confirmation:delete-note'
      confirmMessage='i18n:highlighting:button:delete'
      onConfirm={() => saveAnnotation(assertDefined(props.data, 'props.data should be defined'))}
      onCancel={() => {
        setEditing(true);
        setPendingAnnotation(defaultAnnotation());
      }}
      always={() => setConfirmingDelete(false)}
    />}
    {!authenticated && <Confirmation
      data-analytics-label='login'
      data-analytics-region='highlighting-login'
      message='i18n:highlighting:login:prompt'
      confirmMessage='i18n:highlighting:login:link'
      confirmLink={loginLink}
      onCancel={props.onBlur}
    />}
  </form>;
});

export default styled(EditCard)`
  background: ${theme.color.neutral.formBackground};
  user-select: none;
  overflow: visible;

  ${ButtonGroup} {
    margin-top: ${cardPadding}rem;
  }

  ${theme.breakpoints.touchDeviceQuery(css`
    visibility: hidden;
  `)}
`;
