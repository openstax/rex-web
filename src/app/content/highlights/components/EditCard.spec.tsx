import { Highlight } from '@openstax/highlighter';
import { HighlightUpdateColorEnum } from '@openstax/highlighter/dist/api';
import React, { ReactElement } from 'react';
import renderer from 'react-test-renderer';
import createTestServices from '../../../../test/createTestServices';
import createTestStore from '../../../../test/createTestStore';
import createMockHighlight from '../../../../test/mocks/highlight';
import { testAccountsUser } from '../../../../test/mocks/userLoader';
import { makeFindByTestId } from '../../../../test/reactutils';
import TestContainer from '../../../../test/TestContainer';
import * as selectAuth from '../../../auth/selectors';
import { formatUser } from '../../../auth/utils';
import { assertDocument } from '../../../utils';
import { highlightStyles } from '../../constants';
import { updateHighlight } from '../actions';
import ColorPicker from './ColorPicker';
import EditCard, { EditCardProps } from './EditCard';
import Note from './Note';
import * as onClickOutsideModule from './utils/onClickOutside';

jest.mock('./ColorPicker', () => (props: any) => <div mock-color-picker {...props} />);
jest.mock('./Note', () => (props: any) => <div mock-note {...props} ref={props.textareaRef} />);
jest.mock('./Confirmation', () => (props: any) => <div mock-confirmation {...props} />);

describe('EditCard', () => {
  const highlight = createMockHighlight('asdf');
  const highlightData = highlight.serialize().data;
  const services = createTestServices();
  const store = createTestStore();
  const dispatch = jest.spyOn(store, 'dispatch');
  let editCardProps: Partial<EditCardProps>;

  beforeEach(() => {
    jest.resetAllMocks();
    highlight.elements = [assertDocument().createElement('span')];
    editCardProps = {
      highlight: highlight as unknown as Highlight,
      onBlur: jest.fn(),
      onCancel: jest.fn(),
      onCreate: jest.fn(),
      onRemove: jest.fn(),
      setAnnotationChangesPending: jest.fn(),
    };
  });

  it('matches snapshot when focused', () => {
    const data = {
      color: highlightStyles[0].label,
      ...highlightData,
    };
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard {...editCardProps} data={data} isActive={true} />
      </TestContainer>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot with data', () => {
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard {...editCardProps} isActive={true} />
      </TestContainer>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot when editing', () => {
    highlight.getStyle.mockReturnValue('red');
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          data={highlightData}
          isActive={true}
        />
      </TestContainer>
    );

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('asdf');
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot without data', () => {
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard {...editCardProps} isActive={true} />
      </TestContainer>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('chains ColorPicker onRemove', () => {
    const data = {
      ...highlightData,
      annotation: '',
    };
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          isActive={true}
          data={data}
        />
      </TestContainer>
    );

    const picker = component.root.findByType(ColorPicker);
    renderer.act(() => {
      picker.props.onRemove();
    });

    expect(editCardProps.onRemove).toHaveBeenCalled();
  });

  it('doesn\'t chain ColorPicker onRemove if there is a note', () => {
    const data = {
      ...highlightData,
      annotation: 'asdf',
    };
    const component = renderer.create(
      <TestContainer services={services} store={store}>
            <EditCard {...editCardProps} data={data} isActive={true} />
      </TestContainer>
    );

    const picker = component.root.findByType(ColorPicker);
    renderer.act(() => {
      picker.props.onRemove();
    });

    expect(editCardProps.onRemove).not.toHaveBeenCalled();
  });

  it('doesn\'t chain ColorPicker onRemove if there is a pending note', () => {
    highlight.getStyle.mockReturnValue('red');
    const data = {
      ...highlightData,
      annotation: '',
    };
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          data={data}
          isActive={true}
        />
      </TestContainer>
    );

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('asdf');
    });

    const picker = component.root.findByType(ColorPicker);
    renderer.act(() => {
      picker.props.onRemove();
    });

    expect(editCardProps.onRemove).not.toHaveBeenCalled();
  });

  it('cancelling resets the form state', () => {
    highlight.getStyle.mockReturnValue('red');
    const data = {
      ...highlightData,
      annotation: 'qwer',
    };
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          data={data}
          isActive={true}
        />
      </TestContainer>
    );
    const findByTestId = makeFindByTestId(component.root);

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('asdf');
    });

    expect(component.root.findAllByType('button').length).toBe(2);
    expect(note.props.note).toBe('asdf');

    const cancel = findByTestId('cancel');
    renderer.act(() => {
      cancel.props.onClick({preventDefault: jest.fn()});
    });

    expect(note.props.note).toBe('qwer');
    expect(editCardProps.onBlur).not.toHaveBeenCalled();
    expect(component.root.findAllByType('button').length).toBe(0);
  });

  it('save saves', () => {
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          data={highlightData}
          locationFilterId='locationId'
          pageId='pageId'
          isActive={true}
          onCreate={jest.fn()}
        />
      </TestContainer>
    );
    const findByTestId = makeFindByTestId(component.root);

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('asdf');
    });

    const saveButton = findByTestId('save');
    renderer.act(() => {
      saveButton.props.onClick({preventDefault: jest.fn()});
    });

    expect(dispatch).toHaveBeenCalledWith(updateHighlight({
      highlight: {color: highlightData.style as any, annotation: 'asdf'},
      id: highlightData.id,
    }, {
      locationFilterId: 'locationId',
      pageId: 'pageId',
      preUpdateData: {
        highlight: {
          annotation: highlightData.annotation,
          color: highlightData.style as HighlightUpdateColorEnum,
        },
        id: highlightData.id,
      },
    }));
    expect(editCardProps.onBlur).not.toHaveBeenCalled();
    expect(component.root.findAllByType('button').length).toBe(0);
    expect(editCardProps.onCancel).toHaveBeenCalled();
  });

  it('removing note shows confirmation', () => {
    const data = {
      ...highlightData,
      annotation: 'qwer',
    };
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          data={data}
          isActive={true}
        />
      </TestContainer>
    );
    const findByTestId = makeFindByTestId(component.root);

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('');
    });

    const saveButton = findByTestId('save');
    renderer.act(() => {
      saveButton.props.onClick({preventDefault: jest.fn()});
    });

    expect(() => findByTestId('confirm-delete')).not.toThrow();
  });

  it('confirmation can save', () => {
    const data = {
      ...highlightData,
      annotation: 'qwer',
    };
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          locationFilterId='locationId'
          pageId='pageId'
          isActive={true}
          data={data}
        />
      </TestContainer>
    );
    const findByTestId = makeFindByTestId(component.root);

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('');
    });

    const saveButton = findByTestId('save');
    renderer.act(() => {
      saveButton.props.onClick({preventDefault: jest.fn()});
    });

    const confirmation = findByTestId('confirm-delete');
    renderer.act(() => {
      confirmation.props.onConfirm();
      confirmation.props.always();
    });

    expect(() => findByTestId('confirm-delete')).toThrow();
    expect(dispatch).toHaveBeenCalledWith(updateHighlight({
      highlight: {color: highlightData.style as any, annotation: ''},
      id: highlightData.id,
    }, {
      locationFilterId: 'locationId',
      pageId: 'pageId',
      preUpdateData: {
        highlight: {
          annotation: data.annotation,
          color: data.style as HighlightUpdateColorEnum,
        },
        id: highlightData.id,
      },
    }));
    expect(editCardProps.onBlur).not.toHaveBeenCalled();
    expect(editCardProps.onCancel).toHaveBeenCalled();
  });

  it('confirmation can cancel', () => {
    highlight.getStyle.mockReturnValue('red');
    const data = {
      ...highlightData,
      annotation: 'qwer',
    };
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          data={data}
          isActive={true}
        />
      </TestContainer>
    );
    const findByTestId = makeFindByTestId(component.root);

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('');
    });

    const saveButton = findByTestId('save');
    renderer.act(() => {
      saveButton.props.onClick({preventDefault: jest.fn()});
    });

    const confirmation = findByTestId('confirm-delete');
    renderer.act(() => {
      confirmation.props.onCancel();
      confirmation.props.always();
    });

    expect(() => findByTestId('confirm-delete')).toThrow();
    expect(dispatch).not.toHaveBeenCalled();
    expect(note.props.note).toBe('qwer');
  });

  it('responds to changes', () => {
    highlight.getStyle.mockReturnValue('red');
    const data = {
      ...highlightData,
      annotation: 'qwer',
    };
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          data={data}
          isActive={true}
          hasUnsavedHighlight={false}
        />
      </TestContainer>
    );
    const note = component.root.findByType(Note);

    renderer.act(() => {
      note.props.onChange('');
    });
    expect(editCardProps.setAnnotationChangesPending).toHaveBeenCalledWith(true);
  });

  it('dispatches if changes are reverted', () => {
    highlight.getStyle.mockReturnValue('red');
    const data = {
      ...highlightData,
      annotation: 'qwer',
    };
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          data={data}
          isActive={true}
          hasUnsavedHighlight={true}
        />
      </TestContainer>
    );
    const note = component.root.findByType(Note);

    renderer.act(() => {
      note.props.onChange('qwer');
    });
    expect(editCardProps.setAnnotationChangesPending).toHaveBeenCalledWith(false);
  });

  it('handles color change when there is data', () => {
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          data={highlightData}
          locationFilterId='locationId'
          pageId='pageId'
          isActive={true}
        />
      </TestContainer>
    );

    const picker = component.root.findByType(ColorPicker);
    renderer.act(() => {
      picker.props.onChange('blue');
    });

    expect(highlight.setStyle).toHaveBeenCalledWith('blue');
    expect(store.dispatch).toHaveBeenCalledWith(updateHighlight({
      highlight: {annotation: highlightData.annotation, color: 'blue' as any},
      id: highlightData.id,
    }, {
      locationFilterId: 'locationId',
      pageId: 'pageId',
      preUpdateData: {
        highlight: {
          annotation: highlightData.annotation,
          color: highlightData.style as HighlightUpdateColorEnum,
        },
        id: highlightData.id,
      },
    }));
  });

  it('creates when changing color on a new highlight', () => {
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard {...editCardProps} isActive={true} />
      </TestContainer>
    );

    const picker = component.root.findByType(ColorPicker);
    renderer.act(() => {
      picker.props.onChange('blue');
    });

    expect(highlight.setStyle).toHaveBeenCalledWith('blue');
    expect(editCardProps.onCreate).toHaveBeenCalled();
  });

  it('sets color and creates when you focus', () => {
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard {...editCardProps} />
      </TestContainer>
    );

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onFocus();
    });

    expect(highlight.setStyle).toHaveBeenCalledWith(highlightStyles[0].label);
    expect(editCardProps.onCreate).toHaveBeenCalled();
  });

  it('focusing an existing note does nothing', () => {
    highlight.getStyle.mockReturnValue('red');
    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard {...editCardProps} data={highlightData} />
      </TestContainer>
    );

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onFocus();
    });

    expect(highlight.setStyle).not.toHaveBeenCalled();
    expect(editCardProps.onCreate).not.toHaveBeenCalled();
  });

  it('blurs when clicking outside', () => {
    const onClickOutside = jest.spyOn(onClickOutsideModule, 'useOnClickOutside');
    onClickOutside.mockReturnValue();

    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard {...editCardProps} isActive={true} />
      </TestContainer>
    );

    onClickOutside.mock.calls[0][2]({} as any);

    expect(component).toBeTruthy();
    expect(onClickOutside.mock.calls.length).toBe(1);
    expect(editCardProps.onBlur).toHaveBeenCalled();
  });

  it('doesn\'t blur when clicking outside and editing', () => {
    highlight.getStyle.mockReturnValue('red');

    const onClickOutside = jest.spyOn(onClickOutsideModule, 'useOnClickOutside');
    onClickOutside.mockReturnValue();

    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          isActive={true}
          data={highlightData}
          hasUnsavedHighlight={true}
        />
      </TestContainer>
    );

    renderer.act(() => undefined);

    const note = component.root.findByType(Note);
    renderer.act(() => {
      note.props.onChange('asdf');
    });

    onClickOutside.mock.calls[1][2]({} as any);

    expect(onClickOutside.mock.calls.length).toBe(2);
    expect(editCardProps.onBlur).not.toHaveBeenCalled();
    expect(editCardProps.onCancel).not.toHaveBeenCalled();
  });

  it('trackShowCreate for authenticated user', () => {
    const onClickOutside = jest.spyOn(onClickOutsideModule, 'default');
    onClickOutside.mockReturnValue(() => () => null);

    const mockSpyUser = jest.spyOn(selectAuth, 'user')
      .mockReturnValue(formatUser(testAccountsUser));

    const spyAnalytics = jest.spyOn(services.analytics.showCreate, 'track');

    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          isActive={true}
          data={undefined}
        />
      </TestContainer>
    );

    expect(spyAnalytics).not.toHaveBeenCalled();

    // Wait for React.useEffect
    renderer.act(() => undefined);

    expect(() => component.root.findByProps({
      'data-analytics-region': 'highlighting-login',
    })).toThrow();
    expect(spyAnalytics).toHaveBeenCalled();
    mockSpyUser.mockClear();
  });

  it('call onHeightChange when element mounts', () => {
    const onClickOutside = jest.spyOn(onClickOutsideModule, 'default');
    onClickOutside.mockReturnValue(() => () => null);

    const onHeightChange = jest.fn();
    const createNodeMock = () => assertDocument().createElement('div');

    const component = renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard
          {...editCardProps}
          isActive={true}
          onHeightChange={onHeightChange}
        />
      </TestContainer>,
      {createNodeMock}
    );

    expect(onHeightChange).not.toHaveBeenCalled();

    // Wait for mount
    renderer.act(() => undefined);

    expect(() => component.root.findByProps({
      'data-analytics-region': 'edit-note',
    })).not.toThrow();
    expect(onHeightChange).toHaveBeenCalled();
  });

  it('focuses textarea if shouldFocusCard is set to true', () => {
    const editCard = assertDocument().createElement('div');
    const textarea = assertDocument().createElement('textarea');

    const spyTextareaFocus = jest.spyOn(textarea, 'focus');

    const createNodeMock = (element: ReactElement) => {
      if (element.type === 'form') {
        return editCard;
      } else {
        return textarea;
      }
    };

    const onHeightChange = jest.fn();

    renderer.create(
      <TestContainer services={services} store={store}>
        <EditCard {...editCardProps} onHeightChange={onHeightChange} shouldFocusCard={true} />
      </TestContainer>,
      {createNodeMock}
    );

    // Wait for hooks
    renderer.act(() => undefined);

    expect(spyTextareaFocus).toHaveBeenCalledTimes(1);
  });
});
