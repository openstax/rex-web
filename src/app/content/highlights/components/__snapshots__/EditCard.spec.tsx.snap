// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditCard matches snapshot when editing 1`] = `
<form
  className="EditCard-sc-1kvvj32-0 iFPiln"
  data-analytics-region="edit-note"
  data-highlight-card={true}
>
  <div
    mock-color-picker={true}
    onChange={[Function]}
    onRemove={[Function]}
  />
  <div
    mock-note={true}
    note="asdf"
    onChange={[Function]}
    onFocus={[Function]}
    textareaRef={
      Object {
        "current": null,
      }
    }
  />
  <div
    className="Button__ButtonGroup-ayg7nk-1 fJPuTt"
  >
    <button
      className="Button-ayg7nk-0 dIvXiu"
      data-analytics-label="save"
      data-testid="save"
      onClick={[Function]}
    >
      Save
    </button>
    <button
      className="Button-ayg7nk-0 bHTkEr"
      data-analytics-label="cancel"
      data-testid="cancel"
      onClick={[Function]}
    >
      Cancel
    </button>
  </div>
  <div
    confirmLink="/accounts/login?r=/"
    confirmMessage="i18n:highlighting:login:link"
    data-analytics-label="login"
    data-analytics-region="highlighting-login"
    message="i18n:highlighting:login:prompt"
    mock-confirmation={true}
    onCancel={[MockFunction]}
  />
</form>
`;

exports[`EditCard matches snapshot when focused 1`] = `
<form
  className="EditCard-sc-1kvvj32-0 iFPiln"
  data-analytics-region="edit-note"
  data-highlight-card={true}
>
  <div
    color="yellow"
    mock-color-picker={true}
    onChange={[Function]}
    onRemove={[Function]}
  />
  <div
    mock-note={true}
    note=""
    onChange={[Function]}
    onFocus={[Function]}
    textareaRef={
      Object {
        "current": null,
      }
    }
  />
  <div
    confirmLink="/accounts/login?r=/"
    confirmMessage="i18n:highlighting:login:link"
    data-analytics-label="login"
    data-analytics-region="highlighting-login"
    message="i18n:highlighting:login:prompt"
    mock-confirmation={true}
    onCancel={[MockFunction]}
  />
</form>
`;

exports[`EditCard matches snapshot with data 1`] = `
<form
  className="EditCard-sc-1kvvj32-0 iFPiln"
  data-analytics-region="edit-note"
  data-highlight-card={true}
>
  <div
    mock-color-picker={true}
    onChange={[Function]}
    onRemove={[Function]}
  />
  <div
    mock-note={true}
    note=""
    onChange={[Function]}
    onFocus={[Function]}
    textareaRef={
      Object {
        "current": null,
      }
    }
  />
  <div
    confirmLink="/accounts/login?r=/"
    confirmMessage="i18n:highlighting:login:link"
    data-analytics-label="login"
    data-analytics-region="highlighting-login"
    message="i18n:highlighting:login:prompt"
    mock-confirmation={true}
    onCancel={[MockFunction]}
  />
</form>
`;

exports[`EditCard matches snapshot without data 1`] = `
<form
  className="EditCard-sc-1kvvj32-0 iFPiln"
  data-analytics-region="edit-note"
  data-highlight-card={true}
>
  <div
    mock-color-picker={true}
    onChange={[Function]}
    onRemove={[Function]}
  />
  <div
    mock-note={true}
    note=""
    onChange={[Function]}
    onFocus={[Function]}
    textareaRef={
      Object {
        "current": null,
      }
    }
  />
  <div
    confirmLink="/accounts/login?r=/"
    confirmMessage="i18n:highlighting:login:link"
    data-analytics-label="login"
    data-analytics-region="highlighting-login"
    message="i18n:highlighting:login:prompt"
    mock-confirmation={true}
    onCancel={[MockFunction]}
  />
</form>
`;
